agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1394,21 +1398,35 @@ func (dg *dockerGoClient) Stats(ctx context.Context, id string, inactivityTimeou
 	return statsC, errC
 }
 
-func getContainerStatsNotStreamed(client sdkclient.Client, ctx context.Context, id string) (*types.StatsJSON, error) {
-	resp, err := client.ContainerStats(ctx, id, false)
-	if err != nil {
-		return nil, fmt.Errorf("DockerGoClient: Unable to retrieve stats for container %s: %v", id, err)
+func getContainerStatsNotStreamed(client sdkclient.Client, ctx context.Context, id string, timeout time.Duration) (*types.StatsJSON, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	type statsResponse struct {
+		stats types.ContainerStats
+		err   error
 	}
-	defer resp.Body.Close()
-
-	decoder := json.NewDecoder(resp.Body)
-	stats := &types.StatsJSON{}
-	err = decoder.Decode(stats)
-	if err != nil {
-		return nil, fmt.Errorf("DockerGoClient: Unable to decode stats for container %s: %v", id, err)
+	response := make(chan statsResponse, 1)
+	go func() {
+		stats, err := client.ContainerStats(ctxWithTimeout, id, false)
+		response <- statsResponse{stats, err}
+	}()
+	select {
+	case resp := <-response:
+		decoder := json.NewDecoder(resp.stats.Body)
+		stats := &types.StatsJSON{}
+		err := decoder.Decode(stats)
+		if err != nil {
+			return nil, fmt.Errorf("DockerGoClient: Unable to decode stats for container %s: %v", id, err)
+		}
+		defer resp.stats.Body.Close()
+		return stats, nil
+	case <-ctxWithTimeout.Done():
+		err := ctxWithTimeout.Err()
+		if err == context.DeadlineExceeded {
+			return nil, fmt.Errorf("DockerGoClient: timed out retrieving stats for container %s", id)
+		}
```

Comment:
maybe outside scope of this change, but relatedly I was wondering if we might want to catch and ignore `context.Canceled` errors here? I believe we end up logging these as a warning or error, but they are actually expected to happen when the container exits.

Not sure if there might be some other implication to ignoring these errors though...

Reply:
I think its better to just log them since context cancel could also happen for reasons other than container exit (we log them for other docker APIs too)